### PR TITLE
Do NOT merge: Example PR for implementing list/watch cache

### PR DESCRIPTION
**Pull request by brendandburns**
This shows how we might implement a generic list/watch cache that could be used as a foundation for a generic informer.

### CODE REVIEW

File: util/src/main/java/io/kubernetes/client/util/Cache.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    } catch (NoSuchFieldException | IllegalAccessException ex) {
+      log.error("Error adding item", ex);
+    }
+  }
+
+  private void startWatch() throws ApiException {
+    System.out.println("Starting watch.");
+    Call call = callGenerator.apply(Boolean.TRUE);
+    watch = Watch.createWatch(client, call, new TypeToken<Watch.Response<ApiType>>() {}.getType());
+    watchThread =
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/util/Cache.java]: nit: do we need to set it to daemon?

File: util/src/main/java/io/kubernetes/client/util/Cache.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      throw new ApiException(
+          response.message(), response.code(), response.headers().toMultimap(), respBody);
+    }
+    list = client.getJSON().getGson().fromJson(response.body().charStream(), listClass);
+  }
+
+  private ApiType getItem(String namespace, String name)
+      throws NoSuchFieldException, IllegalAccessException {
+    try {
+      Field itemsField = listClass.getClass().getDeclaredField("items");
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/util/Cache.java]: > it's always a private field, isn't it?

this comment also applies for here

File: util/src/main/java/io/kubernetes/client/util/Cache.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.List;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Cache<ApiType, ApiListType> implements Closeable {
+  private Function<Boolean, Call> callGenerator;
+  private ApiClient client;
+  private Watch<ApiType> watch;
+  private ApiListType list;
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/util/Cache.java]: we still need to index the list, or a namespaced list from a all-namespace cache will be costy

File: util/src/main/java/io/kubernetes/client/util/Cache.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+          return;
+        }
+      }
+    } catch (NoSuchFieldException | IllegalAccessException ex) {
+      log.error("Error adding item", ex);
+    }
+  }
+
+  private void startWatch() throws ApiException {
+    System.out.println("Starting watch.");
```
[the remark this answers is not in the document]
Comment (brendandburns) [reply util/src/main/java/io/kubernetes/client/util/Cache.java]: This is debug code, will remove if we want to take this PR forward.

File: util/src/main/java/io/kubernetes/client/util/Reflect.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package io.kubernetes.client.util;
+
+import io.kubernetes.client.models.V1ObjectMeta;
+import java.lang.reflect.Field;
+
+public class Reflect {
+  public static V1ObjectMeta objectMetadata(Object obj)
+      throws NoSuchFieldException, IllegalAccessException {
+    Field mdField = obj.getClass().getDeclaredField("metadata");
```
[the remark this answers is not in the document]
Comment (brendandburns) [reply util/src/main/java/io/kubernetes/client/util/Reflect.java]: yep, now fixed :)

File: util/src/main/java/io/kubernetes/client/util/Cache.java (hunk 2 above)
Comment (brendandburns) [reply util/src/main/java/io/kubernetes/client/util/Cache.java]: yep fixed now.

File: util/src/main/java/io/kubernetes/client/util/Cache.java (hunk 3 above)
Comment (brendandburns) [reply util/src/main/java/io/kubernetes/client/util/Cache.java]: yep.  There are lots of optimizations we can/should do.

**Comment by brendandburns** [comment]
Ok @yue9944882 this is ready for you to take it over, when you're ready file a new PR and I will close this one.

**Comment by brendandburns** [comment]
Closing in favor of #469

**Pull request closed without merging**